### ISSUE TITLE
linter.unusedVariables prevents using named arguments

### ISSUE THREAD

**Issue by eric-wieser** [leanprover/lean4#3111]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Check that your issue is not already [filed](https://github.com/leanprover/lean4/issues).
    * Reduce the issue to a minimal, self-contained, reproducible test case. Avoid dependencies to mathlib4 or std4.

### Description

The `unusedVariables` linter encourages the user to make changes that break downstream code that uses named arguments `(name := _)` syntax.

### Steps to Reproduce
Run the following:
```lean
theorem zero_lt_length_of_ne_nil :
    ∀ (l : List α) (h : l ≠ []), 0 < l.length
  | _ :: _, _ => Nat.zero_lt_succ _

#check zero_lt_length_of_ne_nil (l := [1]) (h := by simp)
```
**Expected behavior:** No lint errors are reported, as `h` is part of the public API of `zero_lt_length_of_ne_nil`

**Actual behavior:** `unused variable 'h'`. The linter error cannot be satisfied without breaking the `#check` If you replace `h` with `_h`, then the `#check` breaks.

### Versions

"4.5.0-rc1" (web editor)

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others are impacted by this issue, please ask them to add :+1: to it.

**Comment by PatrickMassot** [comment]
It would be nice to have a more precise linter, but honestly I feel the issue in the given example is really wanting to save a couple of characters. You could write
```lean
theorem zero_lt_length_of_ne_nil (l : List α) (h : l ≠ []) : 0 < l.length :=
  match l,h with 
  | _ :: _, _ => Nat.zero_lt_succ _
```
and have your named assumption and no linter complain.

**Comment by eric-wieser** [comment]
Indeed, the error goes away if you write the theorem in that style; but I think this implicit `match` style is intended to be encouraged. Perhaps I should have given a better example, to make it clear that this isn't just a problem with `theorems`:
```lean
def concatBit : ∀ (msbs : List Bool) (lsb : Bool), List Bool
  | [], lsb => [lsb]
  | msb :: msbs, lsb => [msb] ++ concatBit (msbs := msbs) (lsb := lsb)
```
Obviously this is a stupid implementation, but here the intent is clearly to have a public Api with named data arguments, and the linter is encouraging the user to mangle their names and create a worse API.

**Comment by Kha** [comment]
If we had attributes on parameters, it could be `(@[no_lint] h : ...)`

**Comment by eric-wieser** [comment]
That would certainly be a nice feature, but ideally it wouldn't be necessary for this specific case.

**Comment by nomeata** [comment]
I'd agree with Eric that such a variable binding isn't lint. The parameter names are in some sense part of the type and affect the public API, so at least for exported types they can't really be considered “unused”

**Comment by eric-wieser** [comment]
Here's a slightly more tricky example: should this `x` be considered unused? Again, the proof breaks if you only rename the one that is underlined:
```lean
theorem foo (h : ∀ (x : Nat), True) : True := h (x := 1)
```

**Comment by eric-wieser** [comment]
Here's a half-baked proposal that I think would fix this: the linter should only run on lambda `Expr`s not on forall `Expr`s.

**Comment by nomeata** [comment]
It doesn't seem much different? `h : ∀ (x : Nat), True)` and `h : ∀ (_ : Nat), True)` are not interchangeable, and the `x` name clearly is used somewhere in the code. In my understanding it's a linter for unused _names_, not for unused parameters (in the sense that the `Nat` isn't used in `True`).

**Comment by eric-wieser** [comment]
The key difference is that in `(h : ∀ (x : Nat), True)`, the name `h` is part of the public API, but the name `x` is effectively an implementation detail that only matters within the scope of the current declaration. Perhaps that's a distinction we don't care about.

**Comment by digama0** [comment]
Unfortunately almost any binder can end up in downstream code, there is not really a clear boundary for what is accessible. For example: 
```lean
theorem foo (h : ∀ (x : Nat), True) : True := trivial

example : True := by
  suffices _ from
    have _ := foo (h := this)
    this (x := 1)
  exact fun _ => trivial
```
However I think it is reasonable to treat this as an edge case and use the heuristic that variables that appear as direct arguments to the function (i.e. the names of arguments in the forall telescope of the type of the constant) are considered public and others are not.

**Comment by nomeata** [comment]
How important is it to lint unused forall'ed parameters in practice? I.e. how bad would it be to simply not lint there, how many bugs would be prevented? 

(Even for completely “unused” parameters, the name may be useful for readability.)

**Comment by eric-wieser** [comment]
On second thoughts, I think that heuristic is wrong; it would make `∀ (x : Nat), True` lint-free but not `∃ (x : Nat), True`.

Perhaps a better heuristic is "do not apply `unusedVariables` in either `forall`, or `lambda`s whose codomain is a `Type`.

**Comment by nomeata** [comment]
Not sure its wrong; for `∃ (x : Nat), True`, how can that be used in a way that the `x` matters?

**Comment by eric-wieser** [comment]
I guess what I mean is that it feels morally analogous to `∀ (x : Nat), True` (to a mathematician); though you're right that the `x` doesn't actually matter in the same way.

**Comment by nomeata** [comment]
A pedantic language designer might point out that it’s not very principled to use the same name for “internal” use (the one the linter warns about) and “external” use (named arguments), and if that was separate we wouldn’t have this discussion. But such a language design would probably be quite unwieldy and impractical, and anyways we have what we have, so these rough edges around “what is unused” we will have to deal with :-)

**Comment by digama0** [comment]
> A pedantic language designer might point out that it’s not very principled to use the same name for “internal” use (the one the linter warns about) and “external” use (named arguments), and if that was separate we wouldn’t have this discussion. But such a language design would probably be quite unwieldy and impractical,

This is Swift FYI

**Comment by rkirov** [comment]
Just one more real-world example of this type of issue with the unused linter, that I ran into while doing exercises from Tao's Real Analysis companion. The exercises are setup as `Decidable` statements, where the person doing them has to start with `apply isTrue` or `apply isFalse` and complete the proof.

This is my solution for the excerise - https://github.com/rkirov/analysis/blob/main/analysis/Analysis/Section_8_1.lean#L144-L153 and the linter marks `hx` as unused, but it is there for a good reason. Without it the exercise becomes trivially solvable just use empty set.

Here is an even smaller repo in the vain of the exercise above

```lean
def Int.exists_mul : Decidable (∀ (n: Nat) (hn : n ≠ 0), ∃ m, m * n > m) := by
  -- the first line of this construction should be either `apply isTrue` or `apply isFalse`.
  apply isFalse
  simp
  exact ⟨0, by intro h; omega⟩
```

The linters should not compain about `hn` being unused as calling `isFalse` turns it into something to prove and not use (simp proves it).

**Comment by rkirov** [comment]
One more issue I found with the unused linter

```lean
import Mathlib

example : ∃ (X:Type) (h: LE X), (∀ x:X, x ≤ x) := by 
  use Fin 0, ⟨fun _ _ ↦ True⟩; simp
```

https://live.lean-lang.org/#codez=JYWwDg9gTgLgBAWQIYwBYBtgCMBQOCmAHkuOvnAFxyDARHABQAaFAKgJ5j4CU9qVAMgFE4DTgBp6gACI4hCg3GE4gEyJp3CgF44WVnBxw4AVwDO5AGLAAdnAAM4wBfkAM32WA+nFeAywjjMo+/IEvyAG44Q1AwHCA

The linter claims (h: LE X) is unused, but it is needed to be used by `≤` in the next argument.

(let me know if a separate issue for this is preferable)

**Comment by nomeata** [comment]
That's separate, I think, and not actually an issue: `example : ∃ (X:Type) (_: LE X), (∀ x:X, x ≤ x)` works, and the name `h` is really unused.

**Comment by rkirov** [comment]
You are totally right, I keep getting it wrong that in the presence of all implicits and such, "name not used" is different from "hypothesis not used". The linter just checks "name not used", suggesting replacing with `_`, which the hypothesis can very much be still needed as above. Apologies.
